include/grpc/event_engine/posix.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_EVENT_ENGINE_POSIX_H
+#define GRPC_EVENT_ENGINE_POSIX_H
+
+#include <grpc/support/port_platform.h>
+
+#include <functional>
+#include <vector>
+
+#include "absl/functional/any_invocable.h"
+#include "absl/status/status.h"
+
+#include <grpc/event_engine/endpoint_config.h>
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/event_engine/memory_allocator.h>
+#include <grpc/event_engine/port.h>
+#include <grpc/event_engine/slice_buffer.h>
+
+namespace grpc_event_engine {
+namespace experimental {
+
+#ifdef GRPC_EVENT_ENGINE_POSIX
+
+/// This defines an EventEngine interface that all posix specific event engines
+/// must implement.
+class PosixEventEngine : public EventEngine {
+  class PosixEventEngineEndpoint : public EventEngine::Endpoint {
+   public:
+    /// Returns the file descriptor associated with the posix endpoint.
+    virtual int GetWrappedFd() = 0;
+
+    /// Shutdown the endpoint. After this function call its illegal to invoke
+    /// any other methods on the endpoint.
+    /// \a release_fd - a pointer to hold the file descriptor associated with
+    /// the posix endpoint. If it is not nullptr, then the file descriptor is
+    /// not closed while the endpoint is shutdown. The pointer will be updated
+    /// to hold the endpoint's associated file descriptor.
+    /// \a on_release_cb - a callback to be invoked when the endpoint is
+    /// shutdown and the file descriptor is released. It should be invoked only
+    /// if release_fd is not nullptr.
+    virtual void Shutdown(
+        int* release_fd,
```

Comment:
Why not just pass the fd to `on_release_fd()`, so that the caller doesn't have access to the fd until it's actually safe to use?

Reply:
I changed the on_release_fd callback now to take an absl::StatusOr<int> argument (representing the released fd).